src/main/java/reposense/git/GitClone.java

```diff
@@ -41,22 +45,86 @@ public static void clone(RepoConfiguration repoConfig)
+            logger.info("Cloning in parallel from " + repoConfig.getLocation() + "...");
+            Path rootPath = Paths.get(FileUtil.REPOS_ADDRESS, repoConfig.getRepoFolderName());
+            Files.createDirectories(rootPath);
+            crp = runCommandAsync(rootPath, "git clone " + addQuote(repoConfig.getLocation().toString()));
+        } catch (RuntimeException rte) {
+            logger.log(Level.SEVERE, "Error encountered in Git Cloning, will attempt to continue analyzing", rte);
+            throw new GitCloneException(rte);
+        } catch (IOException ioe) {
+            throw new GitCloneException(ioe);
+        }
+    }
+
+    /**
+     * Waits for previously spawned clone process to finish executing.
+     * Should only be called after {@code spawnCloneProcess} has been called.
+     */
+    public static void waitForCloneProcess(RepoConfiguration repoConfig) throws GitCloneException {
+        try {
+            crp.waitForProcess();
+            logger.info("Cloning of " + repoConfig.getLocation() + " completed!");
+        } catch (RuntimeException | CommandRunnerProcessException e) {
+            crp = null;
+            logger.log(Level.SEVERE, "Error encountered in Git Cloning, will attempt to continue analyzing", e);
+            throw new GitCloneException(e);
+        }
+        crp = null;
+        updateRepoConfigBranch(repoConfig);
+    }
 
+    /**
+     * Updates the branch of {@code repoConfig} to the current branch if default branch is specified.
+     */
+    public static void updateRepoConfigBranch(RepoConfiguration repoConfig, String currentBranch) {
+        if (repoConfig.getBranch().equals(RepoConfiguration.DEFAULT_BRANCH)) {
+            repoConfig.setBranch(currentBranch);
+        }
+    }
+
+    private static void updateRepoConfigBranch(RepoConfiguration repoConfig) throws GitCloneException {
         try {
             if (repoConfig.getBranch().equals(RepoConfiguration.DEFAULT_BRANCH)) {
                 String currentBranch = GitBranch.getCurrentBranch(repoConfig.getRepoRoot());
                 repoConfig.setBranch(currentBranch);
             }
-            GitCheckout.checkout(repoConfig.getRepoRoot(), repoConfig.getBranch());
         } catch (RuntimeException e) {
-            logger.log(Level.SEVERE, "Branch does not exist! Analyze terminated.", e);
+            logger.log(Level.SEVERE, "Branch does not exist! Analysis terminated.", e);
             throw new GitCloneException(e);
         }
     }
 
-    private static void clone(RepoLocation location, String repoName) throws IOException {
-        Path rootPath = Paths.get(FileUtil.REPOS_ADDRESS, repoName);
-        Files.createDirectories(rootPath);
-        runCommand(rootPath, "git clone " + addQuote(location.toString()));
+    /**
+     * Checks out the branch specified in {@code repoConfig}.
+     */
+    public static void checkoutBranch(RepoConfiguration repoConfig) throws GitCloneException {
```

Comment:
This should go under `GitCheckout`

Reply:
This is actually a "wrapper" around `GitCheckout.checkout()`, to catch `RuntimeException` and throw `GitCloneException`. Might be weird if I shift this and `GitCheckout` throws `GitCloneException`.

But I do agree its weird to have a function called `checkoutBranch` under `GitClone`. Will think about how to improve it.